pkg/build/api/validation/validation.go

```diff
@@ -149,52 +138,25 @@ func hasProxy(source *buildapi.GitBuildSource) bool {
 	return len(source.HTTPProxy) > 0 || len(source.HTTPSProxy) > 0
 }
 
-func validateSource(input *buildapi.BuildSource) fielderrors.ValidationErrorList {
+func validateSource(input *buildapi.BuildSource, isCustomStrategy bool) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
-	switch input.Type {
-	case buildapi.BuildSourceGit:
-		if input.Git == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("git"))
-		} else {
-			allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
-		}
-		if input.Binary != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when type is Git"))
-		}
-	case buildapi.BuildSourceBinary:
-		if input.Binary == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("binary"))
-		} else {
-			allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
-		}
-		if input.Git != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when type is Binary"))
-		}
-	case buildapi.BuildSourceDockerfile:
-		if input.Dockerfile == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("dockerfile"))
-		} else {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+
+	if input.Git != nil {
+		// git and binary input sources are mutually exclusive
+		if input.Binary != nil && !isCustomStrategy {
```

Comment:
Do we really need Git and Binary to be mutually exclusive?

IMHO this adds cognitive load to the end user. I'd rather document:

> ... these are all possible source types: a, b, c, d, ...
> The strategy determines how each source type is used if multiple are given.
> Source strategy: at least one source type is required...
> Docker strategy: at least one source type is required, and this source type must provide a Dockerfile (either a regular file or inline) [note: all current source types qualify for that]
> Custom strategy: no source is required, and details depend on your builder image.


Than:

> Git and Binary are mutually exclusive, unless strategy is custom.
> ...

I think the case by case validation require us to present a table of what is valid and what is not in the docs, and require users to understand the nuances.